### PR TITLE
Expose solver enums through Smt-Switch API

### PR DESCRIPTION
**Pull request by makaimann**
Tag every solver with an enum so we can identify the backend solver that is used by an SmtSolver pointer. It also provides some utility functions for SolverEnums.

Note: there is not an easy way to expose available_solvers through the API (i.e. outside of the testing infrastructure). This is because the interface is meant to be completely solver-agnostic. Thus, you could build smt-switch once with all the solvers, but only link solvers as you need them. So, there's no reliable way to tell which solvers are available, because it depends on which libraries have been linked to your program. If it becomes a desired feature, maybe there is a way to register available classes based on linked libraries...I'm not sure.

### CODE REVIEW

**Review by yoni206 — COMMENTED** [review]
looks good. some questions inline. in addition: aren't we missing interpolation is a property of a solver?

File: include/solver.h [hunk 1]
```diff
@@ -34,7 +35,7 @@ namespace smt {
 class AbsSmtSolver
 {
  public:
-  AbsSmtSolver(){};
+  AbsSmtSolver(SolverEnum se) : solver_enum(se){};
```
[the remark this answers is not in the document]
Comment (makaimann) [reply include/solver.h:38]: That's a good question! I don't think so: in fact, no one should ever call AbsSmtSolver directly. It should only be implemented by a backend solver, in which case there should be a SolverEnum for that backend. So, I think it's okay.

File: include/solver.h [hunk 2]
```diff
@@ -356,7 +357,10 @@ class AbsSmtSolver
     throw NotImplementedException("Interpolants are not supported by this solver.");
   }
 
+  SolverEnum get_solver_enum() { return solver_enum; };
+
  protected:
+  SolverEnum solver_enum;  ///< an enum identifying the underlying solver
```
[the remark this answers is not in the document]
Comment (makaimann) [reply include/solver.h:363]: So, I've been really bad about this in the codebase so far, but my intention was to start using doxygen style comments. This was one of the options for variables: https://www.doxygen.nl/manual/docblocks.html

File: include/solver_enums.h [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+namespace smt {
+enum SolverEnum
+{
+  BTOR = 0,
+  CVC4,
+  MSAT,
+  YICES2,
+  // have separate enum for solver wrapped by LoggingSolver with Shadow DAG
+  // test all versions of that also
```
[the remark this answers is not in the document]
Comment (makaimann) [reply include/solver_enums.h]: Good catch! Updated with c749b63abc4ea164874be73c162a03b67dc15c16

**Comment by makaimann** [comment]
That's a good point about interpolation! In acc15a26e72e772f866743b4951b7f61cdda4556 I just added a SolverEnum for the mathsat interpolator. 

I'd just like to keep regular solvers and interpolators separate. So, I could have an interpolation attribute, but then all the regular solvers wouldn't have it and the interpolators would only have that feature. 

For now, I was thinking we don't explicitly need an interpolation SolverAttribute. What do you think?

**Review by yoni206 — APPROVED** [review]
Thanks for the changes and answers.

**Pull request merged**